### DISCUSSION TITLE
XML-RPC api

### DISCUSSION THREAD

**Discussion by JOHLC** [JOHLC/pfSense-FauxAPI-ha#18] [Ideas]
#14 - moved to discussion 

@travisghansen: pfSense does indeed have a legit xmlrpc api. Would you have any interest in migrating this to use that instead of fauxapi? I have fairly extensive experience using it as I’ve built other integrations (ie: https://github.com/travisghansen/kubernetes-pfsense-controller and other non public tools) leveraging it.

@JOHLC Is there something this does better? 
Would it support the same functions? 
Any experience with creating components for Home Assistant? 

In the end, my vision would be for this to be a real component in Home Assistant with config flows and automatic setup etc. but my python skills are non-existent.

Going to close this as it isn't so much a bug or feature request but a discussion. Moved here:

**Comment by travisghansen** [comment]
The legit api just greatly lowers the barrier to entry for many folks as installing the fauxapi can be difficult and carries a certain amount of risk (not that the xml rpc api doesn’t have risks also).

My skills with python are also pretty minimal but I helped quite a bit with the opensprinkler integration.

**Reply by JOHLC** [reply]
Can you give me an example of how I would make a call to this to gather say gateway or interface stats?

**Reply by travisghansen** [reply]
Yeah I’ll send a small snippet of python code using the built-in xmlrpc client as an example a little later today.

**Comment by travisghansen** [comment]
I know this is quite delayed but here a crude sample to give an idea..

```
import os
import xmlrpc.client

host     = os.environ.get("PFSENSE_HOST")
protocol = os.environ.get("PFSENSE_PROTOCOL")
username = os.environ.get("PFSENSE_USERNAME")
password = os.environ.get("PFSENSE_PASSWORD")

url = "{protocol}://{username}:{password}@{host}/xmlrpc.php".format(protocol=protocol, username=username, password=password, host=host)
print(url)
# create a ServerProxy with a URI that doesn't respond to XMLRPC requests
proxy = xmlrpc.client.ServerProxy(url)

print(proxy.pfsense.host_firmware_version(1, 60))

data = proxy.pfsense.backup_config_section(["nat"])
print(data["nat"])

data = proxy.pfsense.backup_config_section(["filter"])
print(data["filter"])

data = proxy.pfsense.backup_config_section(["aliases"])
print(data["aliases"])

data = proxy.pfsense.backup_config_section(["interfaces"])
print(data["interfaces"])

script="""
require_once '/usr/local/www/includes/functions.inc.php';
require_once '/etc/inc/config.inc';
require_once '/etc/inc/pfsense-utils.inc';

global $config;
$toreturn = [
  "ok" => true,
  //"config" => json_encode($config),
  "stats" => json_encode(get_stats()),
  //"messages" => json_encode(get_interface_info("wan")),
"""

for i, interface in enumerate(data["interfaces"].keys()):
  print ("string[{}] = {}".format(i, interface))
  script = script + """
  "interface_stats_{interface}" => json_encode(get_interface_info("{interface}")),
""".format(interface=interface)

script = script + """
];
"""
print(script)
print(proxy.pfsense.exec_php(script))
```

**Comment by travisghansen** [comment]
I've gone ahead and implemented a ton of this (have a bunch of sensor available already, gonna wrap those up and then begin on services and switches (for enabling/disabling firewall rules, daemons, etc). Let me know if you have some interest in trying it out.

**Comment by JOHLC** [comment]
More than interested in trying it out! I have not had very much time to work on this project myself.

**Comment by travisghansen** [comment]
OK, got an initial stab at it. Still some stuff to clean up but it's a decent first step.

https://github.com/travisghansen/hass-pfsense

Give it a try and let me know how it goes.
